File: Lista3/Exec1/src/exec1/Contatos.java
import java.lang.*;

/**
 * Classe que cria a estrutura de contatos
 */
public class Contatos{

    private Pessoa[] pessoas;
    private int num_pessoas;
    private final int TAM = 1000;
    /**
     * Constroi o vetor de contatos
     */
    public Contatos() {
        pessoas =  new Pessoa[TAM];
        num_pessoas = 0;
    }

    /**
     * Cadastra pessoas físicas (se houver como)
     */
    public void cadastrarFisica(){
        
        if(num_pessoas > TAM){
            System.out.println("Agenga cheia");
            return ;
        }
        else{
            try {
                //Faz a leitura das entradas do usuário
                System.out.println("Cadastrar Pessoa Física\nInserir dados:\n\n1. Nome");
                String nome = EntradaTeclado.leString();

                int i = procuraContato(nome);
                if(i != -1){ 
                    System.out.println("** Aviso **\nJá há um registro com esse nome"); 
                    return; 
                }

                System.out.printf("Endereço: ");
                String end = EntradaTeclado.leString();
                System.out.printf("Email: ");
                String email = EntradaTeclado.leString();
                System.out.printf("Aniversário (DD/MM/AAAA): ");
                String bday = EntradaTeclado.leString();
                System.out.printf("Estado civil: ");
                String ciestadual = EntradaTeclado.leString();
                System.out.printf("CPF: ");
                double cpf = EntradaTeclado.leDouble();

                i = procuraContato(cpf);
                if(i != -1) { System.out.println("** Aviso **\nJá há um registro com esse CPF"); return; }

                System.out.printf("Insira o telefone: ");
                Double phone = EntradaTeclado.leDouble();
                pessoas[num_pessoas++] = new PessoaFisica(nome, end, email, bday, ciestadual, cpf, phone);
                //Adiciona nova pessoa no vetor de contatos

            } catch(Exception e) {
                System.out.printf("** Aviso **\nNão foi possível realizar o cadastro\nVerifique os valores de entrada");
            }
        }
    }


    //Função para cadastrar pessoa jurídica
    public void cadastrarJuridica() {
        if(num_pessoas == TAM-1) { //Verifica se há espaço no vetor
            System.out.println("ERRO: Agenda cheio"); 
            return;
        }

        try {
            //Faz as leituras do usuário
            System.out.println("**Pessoa Jurídica");
            System.out.printf("\nInsira o nome: ");
            String nome = EntradaTeclado.leString();

            int i = procuraContato(nome); //Verifica se há um registro c msm nome
            if(i != -1) { System.out.println("** Aviso **\nJá há um registro com esse nome"); return; }

            System.out.printf("Eendereço: ");
            String end = EntradaTeclado.leString();
            System.out.printf("Email: ");
            String email = EntradaTeclado.leString();
            System.out.printf("Razão social: ");
            String rsocial = EntradaTeclado.leString();
            System.out.printf("Inscrição estadual: ");
            String iestadual = EntradaTeclado.leString();
            System.out.printf("CNPJ: ");
            int cnpj = EntradaTeclado.leInt();

            i = procuraContato(cnpj); //Verifica se há registro c msm CNPJ
            if(i != -1) { System.out.println("** Aviso **\nJá há um registro com esse CNPJ"); return; }

            pessoas[num_pessoas++] = new PessoaJuridica(nome, end, email, rsocial, iestadual, cnpj);
            //Adiciona nova pessoa no vetor de contatos
            
        } catch(Exception e) {
            System.out.printf("** Aviso **\nNão foi possível realizar o cadastro\nVerifique os valores de entrada");
        }
    }


    //Função para imprimir um contato específico utilizando a toString
    public void imprimirContato(Pessoa pessoa) {
        if(pessoa instanceof PessoaFisica) {
            System.out.printf("" + (PessoaFisica)pessoa);
        } else {
            System.out.printf("" + (PessoaJuridica)pessoa);
        }
    }


    //Função para imprimir todos os contatos
    public void imprimirContatos() {
        for(int i = 0; i < num_pessoas; i++) {
            if(pessoas[i] == null) { continue; }
            imprimirContato(pessoas[i]);
        }
    }


    //Função para encontrar registro a partir do nome
    public int procuraContato(String nome) {
        for(int i = 0; i < num_pessoas; i++) {
            if(pessoas[i] == null) { continue; }
            if(pessoas[i].getnome().equals(nome)) { return i; }
        }
        return -1;
    }

    //Função para encontrar registro a partir do cpf / cnpj
    public int procuraContato(Double code) {
        for(int i = 0; i < num_pessoas; i++) {
            if(pessoas[i] == null) { continue; }
            if(pessoas[i] instanceof PessoaFisica) { 
                PessoaFisica p = (PessoaFisica)pessoas[i];
                if(p.getCPF() == code) { return i; } 
            } else {
                PessoaJuridica p = (PessoaJuridica)pessoas[i];
                if(p.getCNPJ() == code) { return i; }
            }
        }
        return -1;
    }


    //Função para encontrar pessoa específica
    public void encontrarPessoa() {
        System.out.printf("Insira a chave de busca (Nome, CPF ou CNPJ): ");
        String key = new String();
        int index = -1;

        try {
            key = EntradaTeclado.leString();
        } catch(Exception e) {
            System.out.printf("** Aviso **\nNão foi possível ler a chave de busca\nVerifique os valores de entrada");
        }

        try {
            //Caso seja possível converter a chave para número, faz a busca por CPF / CNPJ
            int code = Integer.parseInt(key); 
            index = procuraContato(code);
        } catch(NumberFormatException e) {
            //Caso contrário, faz a busca por nome
            index = procuraContato(key);
        }

        if(index == -1) { System.out.println("** Aviso **\nPessoa não encontrada\n"); return; }
        Pessoa p = pessoas[index];
        imprimirContato(p);
    }


    //Função para remover pessoa do vetor de contatos
    public void removerPessoa() {
        System.out.printf("Insira a chave de busca (Nome, CPF ou CNPJ): ");
        String key = new String();
        int index = -1;

        try {
            key = EntradaTeclado.leString();
        } catch(Exception e) {
            System.out.printf("** Aviso **\nNão foi possível ler a chave de busca\nVerifique os valores de entrada");
        }

        //Mesma busca que a utilizada anteriormente
        try {
            int code = Integer.parseInt(key);
            index = procuraContato(code);
        } catch(NumberFormatException e) {
            index = procuraContato(key);
        }

        if(index == -1) { System.out.println("** Aviso **\nPessoa não encontrada\n"); return; }
        for(int i = index; i < num_pessoas; i++) { //Traz todas as pessoas do vetor uma posição para frente
            pessoas[i] = pessoas[i+1];
        }
        System.out.println("Pessoa removida com sucesso\n");
        num_pessoas--;
    }
    
}

File: Lista4/exec2/Contatos.py
from Pessoa import *
from PessoaFisica import *
from PessoaJuridica import *

TAM = 1000
class Contatos():
    def __init__(self):
        self.pessoas = list()
        self.num_pessoas = 0
    
    def PFisica(self):
        if self.num_pessoas == (TAM - 1):
            print("AVISO\nAgenda de contatos cheia!")
            return
        
        try:
            print('\033[32m' + "\nCadastrar Pessoa Física\n".upper() + '\033[0;0m')

            nome        = input('nome:')

            if self.procuraContato(nome) != -1:
                return ('\nNome já registrado!')
            else:
                end      =    input('endereco: ')
                email    = input('e-mail: ')
                aniv     = input('aniversario: ')
                ciestado = input('estado civil: ')
                cpf      = input('CPF: ')

                self.pessoas.append(PessoaFisica(nome, cpf, end, aniv, email, ciestado))
                self.num_pessoas += 1

        except Exception as e:
            print("Infelizmente, não foi possível realizar o cadastro")
            print(f'Erro: {e}')

    def PJuridica(self):
        if self.num_pessoas == (TAM - 1):
            print(" AVISO \nAgenda de contatos cheia!")
            return

        try:
            print('\033[32m' + "\nCadastrar Pessoa Juridica\n".upper() + '\033[0;0m')
            nome        = input('nome:')

            if self.procuraContato(nome) != -1:
                return ('\nNome já registrado!')
            else:
                end     = input('endereco: ')
                email   = input('e-mail: ')
                iscr    = input('inscrição estadual: ')
                rsocial = input('razão social: ')
                cnpj    = int(input('CNPJ: '))

                self.pessoas.append(PessoaJuridica(nome, cnpj, end, iscr, email, rsocial))
                self.num_pessoas += 1   

        except Exception as e:
            print("Infelizmente, não foi possível realizar o cadastro")
            print(f'Erro: {e}')


    def imprimirContatos(self):
        for pessoa in self.pessoas:
            if len(self.pessoas) == 0:
                continue
            else:
                print(pessoa)

    def Proc_Rem_Pessoa(self, flag: bool): # procura ou remove a pessoa
        # flag = False, remover
        # flag = True , procurar
        print("Nome, CPF ou CNPJ para busca: ")        
        pos = -1
        try:
            cb = input()
        except:
            print("Erro: {e}")

        try:
            pos = self.procuraContato(int(cb))
        except:
            pos = self.procuraContato(cb)

        if pos == -1:
            print("ERRO: Pessoa, infelizmente, não encontrada")
        elif pos != -1 and flag == True: # procurar
            print(self.pessoas[pos])
        else: # remover
            del self.pessoas[pos]
            self.num_pessoas -= 1
            print(f'\nPessoa {str(pos)} removida com sucesso')

    def procuraContato(self, chave):
        if isinstance(chave, int):
            for i in range(len(self.pessoas)):
                if len(self.pessoas) == 0:
                    continue
                if isinstance(self.pessoas[i], PessoaFisica):
                    if int(self.pessoas[i].getCPF()) == chave: 
                        return i
                else: 
                    if self.pessoas[i].getCNPJ() == chave:
                        return i
            return -1
        else: 
            for i in range(self.num_pessoas):
                if len(self.pessoas) == 0:
                    continue
                else:
                    if self.pessoas[i].getNome() == chave:
                        print('\n')
                        return i
            return -1

    def bubbleSort(self, lista: list, flag: bool):
        for passo in range(len(lista) - 1, 0, -1):
            for k in range(passo):
                if flag == True:
                    if lista[k].getCPF() > lista[k + 1].getCPF():
                        temp = lista[k].getCPF()
                        lista[k].setCPF(lista[k + 1].getCPF())
                        lista[k + 1].setCPF(tempo)
                elif flag == False:
                    if lista[k].getCNPJ() > lista[k + 1].getCNPJ():
                        temp = lista[k].getCNPJ()
                        lista[k].setCNPJ(lista[k + 1].getCNPJ())
                        lista[k + 1].setCNPJ(temp)

    def Ordenar(self):
        Pfisica = list()
        PJuridica = list()
        N1 = 0 # conta pessoas fisicas
        N2 = 0 # conta pessoas juridicas

        for i in range(len(self.pessoas)):
            if len(self.pessoas) == 0:
                continue
            if isinstance(self.pessoas[i], PessoaFisica):
                Pfisica.append(self.pessoas[i])
                N1 += 1
            else:
                PJuridica.append(self.pessoas[i])
                N2 += 1

        self.bubbleSort(Pfisica, True)
        self.bubbleSort(PJuridica, False)

        for i in range(len(Pfisica)):
            self.pessoas[i] = Pfisica[i]
        
        for i in range(len(PJuridica)):
            self.pessoas[i + N1] = PJuridica[i]

        print("Agenda ordenada com sucesso")
        print('\n'.join(map(str, self.pessoas)))